File: src/graph/Rect.c
#include "Rect.h"
#include "log.h"
#include <stdlib.h>
#include <math.h>

float* parsegraph_Rect_new(apr_pool_t* pool)
{
    float* rv;
    if(pool) {
        rv = apr_pcalloc(pool, sizeof(float)*4);
    }
    else {
        rv = calloc(4, sizeof(float));
    }
    return rv;
}

float* parsegraph_createRect(apr_pool_t* pool, float x, float y, float width, float height)
{
    float* rect = parsegraph_Rect_new(pool);
    parsegraph_Rect_set(rect, x, y, width, height);
    return rect;
}

void parsegraph_Rect_isNaN(float* rect)
{
    return isnan(rect->_x) || isnan(rect->_y) || isnan(rect->_width) || isnan(rect->_height);
}

void parsegraph_Rect_toNaN(float* rect)
{
    rect->_x = NAN;
    rect->_y = NAN;
    rect->_width = NAN;
    rect->_height = NAN;
}

void parsegraph_Rect_clear(float* rect)
{
    parsegraph_Rect_toNaN(rect);
}

void parsegraph_Rect_reset(float* rect)
{
    parsegraph_Rect_toNaN(rect);
}

float parsegraph_Rect_x(float* rect)
{
    return rect[0];
}

void parsegraph_Rect_setX(float* rect, float x)
{
    rect[0] = x;
}

float parsegraph_Rect_y(float* rect)
{
    return rect[1];
}

void parsegraph_Rect_setY(float* rect, float y)
{
    rect[1] = y;
}

float* parsegraph_Rect_clone(float* rect, float* target)
{
    if(!target) {
        target = parsegraph_Rect_new(0);
    }
    parsegraph_Rect_copy(rect, target);
    return target;
}

void parsegraph_Rect_copy(float* src, float* dest)
{
    if(!dest) {
        return parsegraph_Rect_clone(src, 0);
    }
    parsegraph_Rect_copyFrom(dest, src);
}

void parsegraph_Rect_copyFrom(float* dest, float* src)
{
    parsegraph_Rect_set(dest, src[0], src[1], src[2], src[3]);
}

void parsegraph_Rect_translate(float* rect, float x, float y)
{
    parsegraph_Rect_setX(rect, parsegraph_Rect_x(rect) + x);
    parsegraph_Rect_setY(rect, parsegraph_Rect_y(rect) + y);
}

void parsegraph_Rect_scale(float* rect, float sx, float sy)
{
    parsegraph_Rect_setX(rect, parsegraph_Rect_x(rect) * sx);
    parsegraph_Rect_setY(rect, parsegraph_Rect_y(rect) * sy);
    parsegraph_Rect_setWidth(rect, parsegraph_Rect_width(rect) * sx);
    parsegraph_Rect_setHeight(rect, parsegraph_Rect_height(rect) * sy);
}

float parsegraph_Rect_height(float* rect)
{
    return rect[3];
}

void parsegraph_Rect_setHeight(float* rect, float height)
{
    rect[3] = height;
}

float parsegraph_Rect_width(float* rect)
{
    return rect[2];
}

void parsegraph_Rect_setWidth(float* rect, float width)
{
    rect[2] = width;
}

void parsegraph_Rect_setIdentity(float* rect)
{
    memset(rect, 0, sizeof(float)*4);
}

void parsegraph_Rect_set(float* rect, float x, float y, float width, float height)
{
    rect[0] = x;
    rect[1] = y;
    rect[2] = width;
    rect[3] = height;
}

void parsegraph_Rect_destroy(float* rect, apr_pool_t* pool)
{
    if(!pool) {
        free(rect);
    }
}

float parsegraph_Rect_vMin(float* rect)
{
    return parsegraph_Rect_y(rect) - parsegraph_Rect_height(rect)/2;
}

float parsegraph_Rect_vMax(float* rect)
{
    return parsegraph_Rect_y(rect) + parsegraph_Rect_height(rect)/2;
}

float parsegraph_Rect_hMin(float* rect)
{
    return parsegraph_Rect_x(rect) - parsegraph_Rect_width(rect)/2;
}

float parsegraph_Rect_hMax(float* rect)
{
    return parsegraph_Rect_x(rect) + parsegraph_Rect_width(rect)/2;
}

void parsegraph_Rect_include(float* rect, float bx, float by, float bwidth, float bheight)
{
    float ax = parsegraph_Rect_x(rect);
    float ay = parsegraph_Rect_y(rect);
    float awidth = parsegraph_Rect_width(rect);
    float aheight = parsegraph_Rect_height(rect);

    float leftEdge = fminf(ax-awidth/2, bx-bwidth/2);
    float rightEdge = fmaxf(ax+awidth/2, bx+bwidth/2);
    float topEdge = fminf(ay-aheight/2, by-bheight/2);
    float bottomEdge = fmaxf(ay+aheight/2, by+bheight/2);

    float w = rightEdge - leftEdge;
    float h = bottomEdge - topEdge;
    float x = leftEdge + w/2;
    float y = topEdge + h/2;
    parsegraph_Rect_set(rect, x, y, w, h);
}

File: src/graph/Rect.ts
import parsegraph_TestSuite from '../TestSuite';

export default class Rect
{
    _x:number;
    _y:number;
    _width:number;
    _height:number;

    constructor(x?:number, y?:number, width?:number, height?:number) {
        this._x = x;
        this._y = y;
        this._width = width;
        this._height = height;
    }

    isNaN():boolean {
        return isNaN(this._x)
            || isNaN(this._y)
            || isNaN(this._width)
            || isNaN(this._height);
    }

    toNaN():void {
        this._x = NaN;
        this._y = NaN;
        this._width = NaN;
        this._height = NaN;
    }

    clear():void {
        this.toNaN();
    }

    reset():void {
        this.toNaN();
    }

    x():number {
        return this._x;
    }

    setX(x:number):void {
        this._x = x;
    }

    y():number {
        return this._y;
    }

    setY(y:number):void {
        this._y = y;
    }

    clone(target?:Rect) {
        if(target) {
            this.copy(target);
            return target;
        }
        return new Rect(this.x(), this.y(), this.width(), this.height());
    }

    copy(dest?:Rect):Rect {
        if(!dest) {
            return this.clone();
        }
        dest.setX(this.x());
        dest.setY(this.y());
        dest.setWidth(this.width());
        dest.setHeight(this.height());
        return dest;
    }

    translate(x:number, y:number):void {
        this.setX(this.x() + x);
        this.setY(this.y() + y);
    }

    scale(sx:number, sy?:number):void {
        if(arguments.length < 2) {
            sy = sx;
        }
        this.setX(this.x() * sx);
        this.setY(this.y() * sy);
        this.setWidth(this.width() * sx);
        this.setHeight(this.height() * sy);
    }

    height():number {
        return this._height;
    }

    setHeight(height:number):void {
        this._height = height;
    }

    width():number {
        return this._width;
    }

    setWidth(width:number):void {
        this._width = width;
    }

    toString():string {
        return "[Rect x=" + this.x() + ", y=" + this.y() + ", w=" + this.width() + ", h=" + this.height() + "]";
    }

    vMin():number {
        return this.y() - this.height()/2;
    }

    vMax():number {
        return this.y() + this.height()/2;
    }

    hMin():number {
        return this.x() - this.width()/2;
    }

    hMax():number {
        return this.x() + this.width()/2;
    }

    include(bx:number, by:number, bwidth:number, bheight:number):void {
        if(this.isNaN()) {
            this._x = bx;
            this._y = bx;
            this._width = bwidth;
            this._height = bheight;
            return;
        }
        var ax = this._x;
        var ay = this._y;
        var awidth = this._width;
        var aheight = this._height;

        var leftEdge = Math.min(ax-awidth/2, bx-bwidth/2);
        var rightEdge = Math.max(ax+awidth/2, bx+bwidth/2);
        var topEdge = Math.min(ay-aheight/2, by-bheight/2);
        var bottomEdge = Math.max(ay+aheight/2, by+bheight/2);

        var w = rightEdge - leftEdge;
        var h = bottomEdge - topEdge;
        var x = leftEdge + w/2;
        var y = topEdge + h/2;

        this._x = x;
        this._y = y;
        this._width = w;
        this._height = h;
    }
}

const parsegraph_Rect_Tests = new parsegraph_TestSuite("parsegraph_Rect");

parsegraph_Rect_Tests.addTest("vMin", function() {
    var r = new Rect(0, 0, 200, 200);
    if(r.vMin() !== -100) {
        return "vMin, expected -100, got " + r.vMin();
    }
});

parsegraph_Rect_Tests.addTest("vMax", function() {
    var r = new Rect(0, 0, 200, 200);
    if(r.vMax() !== 100) {
        return "vMax, expected 100, got " + r.vMax();
    }
});

parsegraph_Rect_Tests.addTest("hMin", function() {
    var r = new Rect(0, 0, 300, 200);
    if(r.hMin() !== -150) {
        return "vMin, expected -150, got " + r.vMin();
    }
});

parsegraph_Rect_Tests.addTest("hMax", function() {
    var r = new Rect(0, 0, 300, 200);
    if(r.hMax() !== 150) {
        return "hMax, expected 150, got " + r.vMax();
    }
});

parsegraph_Rect_Tests.addTest("include", function() {
    var r = new Rect(0, 0, 200, 200);
    r.include(0, 400, 200, 200);

    if(r.vMax() !== new Rect(0, 400, 200, 200).vMax()) {
        return "vMax failed to adjust";
    }
    //console.log(r);
});
